runtime/gc_vlhgc/MemorySubSpaceTarok.cpp

```diff
@@ -1355,18 +1384,169 @@ MM_MemorySubSpaceTarok::calculateExpandSize(MM_EnvironmentBase *env, UDATA bytes
+			sizeChangeFactor = 2;	
+
+		} else if (currentHybridHeapScore <= (double)_extensions->heapContractionGCTimeThreshold) {
+			/* Try to bring hybridHeapScore a little bit above _extensions->heapContractionGCTimeThreshold.
+			 * If this doesn't cause the hybrid heap score to increase enough so that it is within acceptable bounds, this path will be taken 
+			 * again later, and another small contraction will occur.
+			 */
+			percentDiff = currentHybridHeapScore - (double)_extensions->heapContractionGCTimeThreshold;
+			sizeChangeFactor = 1;
+		} 
+
+		/* Since percentDiff is 0 - 100, make sure to convert it to 0-1 percentage */
+		double heapSizePercentChange = (1.0 + ((double)sizeChangeFactor * (percentDiff / 100 )));
+		Trc_MM_MemorySubSpaceTarok_getHeapSizeWithinBounds_2(env->getLanguageVMThread(), heapSizePercentChange);
+		recommendedHeapSize = (uintptr_t)(heapSizePercentChange * recommendedHeapSize);
+	}
+
+	return recommendedHeapSize;
+}
+
+
+double 
+MM_MemorySubSpaceTarok::calculateGcPctForHeapChange(MM_EnvironmentBase *env, intptr_t heapSizeChange)
+{
+	MM_EnvironmentVLHGC *envVLHGC = (MM_EnvironmentVLHGC *)env;
+
+	/* 
+	 * Since we only attempt to resize after PGC's and Global GC's, if the PGC count has increased, it means we are resizing after a PGC
+	 * and the GMP and PGC statistics we have collected, are useable. Otherwise, use backup methods which are independant of heapSizeChange
+	 */
+
+	if (_previouslyObservedPGCCount != envVLHGC->_heapSizingData.pgcCountSinceGMPEnd) {
+		/* Resizing after a PGC */
+
+		uintptr_t pgcCount = envVLHGC->getRepresentativePgcPerGmpCount();
+
+		if ((0 == pgcCount) && (0 == _lastObservedGcPercentage)) {
+			/* Very first time we are resizing, assume GC % is heapExpansionGCTimeThreshold */
+			_lastObservedGcPercentage = (double)_extensions->heapExpansionGCTimeThreshold;
+
+		} else {
+
+			if (0 != heapSizeChange) {
+				/* 
+				 * Determine what the gc cpu % would be if we changed the heap by heapSizeChange bytes
+				 * Main idea is that the change in number of pgc's (per gmp) will be proportional to how much free tenure changes;
+				 */
+				MM_EnvironmentVLHGC *envVLHGC = (MM_EnvironmentVLHGC *)env;
+				uintptr_t currentFreeTenure = (uintptr_t)envVLHGC->_heapSizingData.freeTenure;
+				uintptr_t potentialFreeTenure = currentFreeTenure + heapSizeChange;
+				if (heapSizeChange <= (-1 * (intptr_t)currentFreeTenure) ) {
+					/* If we try to shrink too much, too fast, tenure will be way too small, causing lots of GC work */
+					potentialFreeTenure = 1;
+				}
+				pgcCount = (uintptr_t)(((double)potentialFreeTenure / currentFreeTenure) * pgcCount);
+			}
+
+			double gcActiveTime = (double)(pgcCount * envVLHGC->_heapSizingData.avgPgcTimeUs) + envVLHGC->_heapSizingData.gmpTime; 
+			double gcInterval = (double)(pgcCount  * (envVLHGC->_heapSizingData.avgPgcTimeUs + envVLHGC->_heapSizingData.avgPgcIntervalUs)) + envVLHGC->_heapSizingData.gmpTime;
```

Comment:
don't know what's meaning of gmpTime yet (is it CPU time spent, or wall clock time?)

I'm thinking that gcActiveTime should account for GMP CPU time (so that purely concurrent work would not be accouted, although up for debate), and gcInterval should account for GMP wall clock time

Reply:
gmpTime is cpu time for GMP work (increments + sweep + concurrentWork/2), which does indeed mean that gcActiveTime includes GMP cpu time. 

I agree that the interval should be "wall clock" between start/end of GMP cycles. I have local variable called `globalMarkIntervalTime` in `SchedulingDelegate`, that I could pass over to be used instead of  `_heapSizingData.gmpTime`.
ie:
```
double gcInterval = (double)(pgcCount  * (envVLHGC->_heapSizingData.avgPgcTimeUs + envVLHGC->_heapSizingData.avgPgcIntervalUs)) + envVLHGC->_heapSizingData.gmpIntervalTime;
```

However, this will drastically change the gc % as reported by this function. If the interval between 2 gmp start (or end), is say 10s, and gc active time is 1s, then if we add GMP interval on the bottom (`gcInterval`), we effectively double the interval (10s becomes 20s), while the `gcActiveTime` remains the same. 

If we remove any GMP considerations in the bottom (`gcInterval`), then this is wall clock time between GMP starts - however, if GMP is very expensive, the gc % could be measured at over 100%. 
ex:
pgc count = 3,
pgc time = 20,
pgc interval = 200,
GMP workTime = 700

My original thought about keeping `envVLHGC->_heapSizingData.gmpTime` as part of the `gcInterval ` consideration, is that if `pgcCount` is very very low (say less than 3), then the measured GC percentage would be high, since GMP would constantly be active. On the other hand, if `pgcCount ` is very high, GMP work/time does not really matter into the overall GC %.

Reply:
I was thinking for gcInterval to use wall clock time _only_, without any PGC duration/interval addons. Sure if you start counting both, you'll end up with about double GMP interval values.